refactor: reduce memory usage

flavio opened the pull request:
Reduce the memory consumption of Policy Server when multiple instances of the same Wasm module are loaded.

Thanks to this change, a worker will have only once instance of `PolicyEvaluator` (hence of wasmtime stack), per unique type of module.

Literally speaking, if a user has the `apparmor` policy deployed 5 times (different names, settings,...) only one instance of `PolicyEvaluator` will be allocated for it.

Warning: the optimization works at the worker level. Meaning that `PolicyEvaluator` are NOT sharing these instances between themselves.

This commit helps to address issue https://github.com/kubewarden/kubewarden-controller/issues/528

Note 1: this depends on https://github.com/kubewarden/policy-evaluator/pull/390

## Benchmark data

I've collected data about the amount of memory consumed by 1 instance of policy server. The test has been done using the following policies being loaded.

<details>
<summary>policies.yml</summary>

```yaml
pod-privileged-protect-mode:
  url: ghcr.io/kubewarden/tests/pod-privileged:v0.2.1
  policyMode: protect
  allowedToMutate: false
  settings: {}

pod-privileged-protect-mode2:
  url: ghcr.io/kubewarden/tests/pod-privileged:v0.2.1
  policyMode: protect
  allowedToMutate: false
  settings: {}

pod-privileged-protect-mode3:
  url: ghcr.io/kubewarden/tests/pod-privileged:v0.2.1
  policyMode: protect
  allowedToMutate: false
  settings: {}

pod-privileged-protect-mode4:
  url: ghcr.io/kubewarden/tests/pod-privileged:v0.2.1
  policyMode: protect
  allowedToMutate: false
  settings: {}

pod-privileged-monitor-mode:
  url: ghcr.io/kubewarden/tests/pod-privileged:v0.2.1
  policyMode: monitor
  allowedToMutate: false
  settings: {}

sleep:
  url: ghcr.io/kubewarden/tests/sleeping-policy:v0.1.0
  policyMode: protect
  allowedToMutate: false
  settings:
    sleepMilliseconds: 2000

disallow-service-loadbalancer:
  url: ghcr.io/kubewarden/tests/disallow-service-loadbalancer:v0.1.5
  policyMode: protect
  allowedToMutate: false

flux:
  url: ghcr.io/kubewarden/tests/go-wasi-template:v0.1.0
  policyMode: protect
  allowedToMutate: true
  settings:
    requiredAnnotations:
      "fluxcd.io/cat": "felix"

flux2:
  url: ghcr.io/kubewarden/tests/go-wasi-template:v0.1.0
  policyMode: protect
  allowedToMutate: true
  settings:
    requiredAnnotations:
      "fluxcd.io/cat": "felix"

verify1:
  url: ghcr.io/kubewarden/test/verify-image-signatures:v0.2.8
  policyMode: protect
  allowedToMutate: true
  settings:
    signatures:
      - image: "*"
        githubActions:
          owner: "kubewarden"
          repo: "app-example"

verify2:
  url: ghcr.io/kubewarden/test/verify-image-signatures:v0.2.8
  policyMode: protect
  allowedToMutate: true
  settings:
    signatures:
      - image: "*"
        githubActions:
          owner: "kubewarden"
          repo: "app-example"

verify3:
  url: ghcr.io/kubewarden/test/verify-image-signatures:v0.2.8
  policyMode: protect
  allowedToMutate: true
  settings:
    signatures:
      - image: "*"
        githubActions:
          owner: "kubewarden"
          repo: "app-example"

verify4:
  url: ghcr.io/kubewarden/test/verify-image-signatures:v0.2.8
  policyMode: protect
  allowedToMutate: true
  settings:
    signatures:
      - image: "*"
        githubActions:
          owner: "kubewarden"
          repo: "app-example"
```
</details>

This configuration file will load 13 policies:

* 4 instances of the "verify signatures" policy
* 5 instances of the "pod privileged" policy
* 2 instances of the go-wasi template policy (remember: this is a 20 MB policy!)
* 1 rego policy
* 1 regular rust policy

### Measurement process

The policy server has been started with an increasing number of workers. I started from 1 worker and went up to 8.
For each worker value, 10 samplings have been taken.

The table below shows the data collected.

| workers | Current (MB) | Optimized (MB) | Improvement % | Current StdDev | Optimized StdDev |
|---------|---------|-----------|---------------|----------------|------------------|
| 1       | 621.00  | 530.38    | -14.59        | 30.86          | 23.26            |
| 2       | 807.35  | 586.15    | -27.40        | 39.24          | 18.24            |
| 3       | 983.33  | 649.17    | -33.98        | 23.23          | 30.79            |
| 4       | 1172.28 | 752.59    | -35.80        | 30.85          | 27.50            |
| 5       | 1358.67 | 831.02    | -38.84        | 22.24          | 27.58            |
| 6       | 1535.16 | 914.64    | -40.42        | 30.53          | 21.74            |
| 7       | 1694.57 | 980.81    | -42.12        | 26.48          | 17.79            |
| 8       | 1897.01 | 1054.93   | -44.39        | 27.44          | 22.69            |


This is a chart that shows the overall memory reduction brought by this change:

![memory optimization](https://github.com/kubewarden/policy-server/assets/22728/fb8d6c80-751c-4067-9297-0eb156b4b752)

flavio commented:
I've rebased against the latest changes merged into the `main` branch

fabriziosestito commented on src/workers/worker.rs:
```diff
@@ -200,25 +196,30 @@ impl Worker {
-                    // allowed to mutate.
-
-                    if adm_req.namespace == Some(namespace.to_string()) {
+                // TODO: we should check immediately if the request is coming from the "always
```
we should track this as an issue / tech-debt

flavio replied:
Done with https://github.com/kubewarden/policy-server/issues/599

fabriziosestito reviewed (approved):
LGTM

viccuad reviewed (approved):
LGTM! Top!

flavio commented on src/workers/worker.rs:
```diff
@@ -200,25 +196,30 @@ impl Worker {
+                        .evaluation_environment
+                        .should_always_accept_requests_made_inside_of_namespace(req_namespace)
+                    {
+                        // given namespace, just set the `allowed`
```
I prefer to leave the comment as is, moreover we are going to remove this whole code later when we fix https://github.com/kubewarden/policy-server/issues/599

flavio commented:
I've merged the changes into the `main` branch of `policy-evaluator`, I've updated this PR to consume policy-evaluator from a specific commit of the original repository.

I would prefer to not tag a new version of policy-evaluator right now. @fabriziosestito is going to open a new PR against `policy-evaluator` that will remove the majority of the locks. I would prefer to have his changes merged into the policy-evaluator `main` branch, and **then** tag a new version of the crate

flavio commented:
Moving to blocked, waiting for the next iteration of policy-evaluator that is lock-free before merging this PR.

fabriziosestito commented on e2e-tests/test_data/policies.yaml:
```diff
@@ -77,3 +77,26 @@ raw-mutation-wasi:
+      overwrite: false
+    supplemental_groups:
+      rule: RunAsAny
+      overwrite: false
```
why is this needed? we are going to remove venom tests soonish

flavio replied:
I'm using them inside of k6 tests. This file should definitely be moved to the load-testing repo once we remove venom

fabriziosestito commented on src/workers/evaluation_environment.rs:
```diff
@@ -0,0 +1,333 @@
+}
+
+#[cfg(test)]
+mod tests {
```
maybe we should add more unit tests here if possible,  for instance we could test the public methods that do not involve an evaluation.

flavio replied:
On it :+1:

flavio replied:
Done

fabriziosestito reviewed (approved):
LGTM, just a couple of comments added

flavio commented on src/workers/pool.rs:
```diff
@@ -0,0 +1,352 @@
+lazy_static! {
+    /// Used to create unique worker IDs
+    static ref WORKER_ID: Arc<AtomicU64> = Arc::new(AtomicU64::new(0));
+}
```
Good catch, I've removed it

flavio commented:
@fabriziosestito , @jvanz , @viccuad : you can take another look at the PR. I've applied the changes you requested with the latest commits

viccuad reviewed (approved):
:rocket:

flavio commented:
Before merging this PR I want to summarize what happened.

# Optimization number 1

The PR started with one optimization technique: ensure each worker of Policy Server (where a worker is a thread dedicated to Wasm evaluation) has just one instance of `PolicyEvaluator` (a WebAssembly sandbox) per WebAssembly module. That means that if a user has the same policy defined 5 times, each time with a different configuration, only one `PolicyEvaluator` will be created inside of a worker.
However, that still means that the same `PolicyEvaluator` is duplicated across each worker. Which explains why the memory consumption is directly proportional with the number of workers.

# Optimization number 2

This is an iteration over the previous one. We changed drastically the way we manage WebAssembly runtimes. We rely on [`wasmtime::InstancePre`](https://docs.rs/wasmtime/latest/wasmtime/struct.InstancePre.html) to keep a deduplicated list of all the policies defined by the user.
This collection of instances is shared across all the workers, which leads to huge memory savings.

In addition to that, we changed the evaluation mode to be "on demand". That means that, once an `AdmissionReview` is received, a fresh `PolicyEvaluator` is created. The creation time is significantly reduced by using `wasmtime::InstancePre`.
Once the evaluation is done, the `PolicyEvaluator` instance is discarded. This leads to the deallocation of the memory used by the WebAssembly sandbox wrapped by `PolicyEvaluator`.

This architecture has two main advantages:
* When dealing with policies that leak memory: reduce the impact they have on the long running Policy Server process
* Ensure each evaluation starts with a clean slate. This prevents bugs caused by policies that are mistakenly leaving unclean state in between evaluations

# Benchmarks

I've updated the table and the graph shown when the PR was open to include also the numbers of the second, and final, optimization.

The benchmarks have been done on the same machine, with the same set of policies and the same measurement process.

| workers | Current | Optimized #1 | Optimized #2 | Improvement #1 % | Improvement #2 % | Current StdDev | Optimized #1 StdDev | Optimized #2 StdDev |
|---------|---------|--------------|--------------|------------------|------------------|----------------|---------------------|---------------------|
| 1       | 621.00  | 530.38       | 526.073      | -14.59           | -15.29           | 30.86          | 23.26               | 20.15               |
| 2       | 807.35  | 586.15       | 535.173      | -27.40           | -33.71           | 39.24          | 18.24               | 13.49               |
| 3       | 983.33  | 649.17       | 532.808      | -33.98           | -45.82           | 23.23          | 30.79               | 17.44               |
| 4       | 1172.28 | 752.59       | 529.324      | -35.80           | -54.85           | 30.85          | 27.50               | 21.26               |
| 5       | 1358.67 | 831.02       | 536.239      | -38.84           | -60.53           | 22.24          | 27.58               | 25.76               |
| 6       | 1535.16 | 914.64       | 533.919      | -40.42           | -65.22           | 30.53          | 21.74               | 18.04               |
| 7       | 1694.57 | 980.81       | 524.528      | -42.12           | -69.05           | 26.48          | 17.79               | 19.15               |
| 8       | 1897.01 | 1054.93      | 541.903      | -44.39           | -71.43           | 27.44          | 22.69               | 9.08                |

This is the updated chart:

![policy-server-memory-opt2](https://github.com/kubewarden/policy-server/assets/22728/f76c3232-67c6-4140-acda-c2030e8affe7)

As you can see, the memory consumption has been significantly reduced. Moreover, the memory usage stays constant regardless of the number of workers instantiated.
